components/consumers/pdf/main.go

```diff
@@ -82,68 +85,79 @@ func sendToS3(filename, bucket, region string) error {
 		Body:   bytes.NewReader(data),
 	})
 	if err != nil {
-		return fmt.Errorf("unable to upload %q to %q: %w", filename, bucket, err)
+		return errors.Errorf("unable to upload %s to %s: %w", filename, bucket, err)
 	}
 
-	fmt.Printf("Successfully uploaded %q to %q\n", filename, bucket)
+	pdfFilename := strings.Replace(filename, ".html", "", -1) + ".pdf"
+	_, err = uploader.Upload(&s3manager.UploadInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(pdfFilename),
+		Body:   bytes.NewReader(pdfBytes),
+	})
+	if err != nil {
+		return errors.Errorf("unable to upload %s to %s: %w", filename, bucket, err)
+	}
+
+	slog.Info("uploaded", "filename", filename, "pdf filename", pdfFilename, "to", "bucket", bucket, "successfully")
 	return nil
 }
 
-func buildPdf(data any) (string, error) {
+func buildPdf(data any) (string, []byte, error) {
 	tmpl := template.Must(template.ParseFiles("default.html"))
 	cleanupRun := func(msg string, cleanup func() error) {
 		if err := cleanup(); err != nil {
-			log.Printf(msg, err)
+			slog.Error(msg, "error", err)
 		}
 	}
 
 	currentPath, err := os.Getwd()
 	if err != nil {
-		return "", fmt.Errorf("could not get current working directory: %w", err)
+		return "", nil, errors.Errorf("could not get current working directory: %w", err)
 	}
 
 	reportHTMLPath := filepath.Join(currentPath, "report.html")
 	//#nosec: G304
 	f, err := os.OpenFile(reportHTMLPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0o600) //#nosec: G304
 	if err != nil {
-		return "", fmt.Errorf("could not open report.html: %w", err)
+		return "", nil, errors.Errorf("could not open report.html: %w", err)
 	}
 	defer cleanupRun("could not close file: %w", f.Close)
 
 	if err = tmpl.Execute(f, data); err != nil {
-		return "", fmt.Errorf("could not apply data to template: %w", err)
+		return "", nil, errors.Errorf("could not apply data to template: %w", err)
 	}
 
 	pw, err := playwright.Run()
 	if err != nil {
-		return "", fmt.Errorf("could not launch playwright: %w", err)
+		return "", nil, errors.Errorf("could not launch playwright: %w", err)
 	}
-	defer cleanupRun("could not stop Playwrigh: %w", pw.Stop)
+	defer cleanupRun("could not stop Playwright: %w", pw.Stop)
 
 	browser, err := pw.Chromium.Launch()
 	if err != nil {
-		return "", fmt.Errorf("could not launch Chromium: %w", err)
+		return "", nil, errors.Errorf("could not launch Chromium: %w", err)
 	}
-	defer cleanupRun("could not close browser: %w", browser.Close)
```

Comment:
browser.close is needed when you plan on opening multiple browsers multiple times. we are closing playwright and immediately after that stopping the cotnainer, there is no reason to do anything else

Reply:
under this logic, we shouldn't close file handles, byte streams, connections etc, since everything is running inside a container. but we do clean up things all the time, because you never know how leaving behind all sorts of running resources will interact with other things that you are doing even in short lived workloads. the components are not supposed to have an infinite amount of resources, they should use the minimum things possible and they should keep their environment tidy